### ISSUE TITLE
Internationalization callback

### ISSUE THREAD

**Issue by gtanczyk** [highcharts/highcharts#13167]
# Description of the feature
During internationalization process I have stumbled upon Highcharts translation capabilities. I'm using `setOptions({ lang: ... })` feature to provide translations. Unfortunately this process have two problems:
1. It's not future proof, in case of new strings appearing in new version of Highcharts I will not be able to detect them easily
2. It's not easy to extract subset of translation keys, which are actually used by my application of Highcharts. Actually this subset in my case is very small.

I would like to suggest translation callback, which will allow me to:
1. Detect untranslated keys from Highcharts, which are actually in use.
2. Integrate with external i18n solution, for example `t` function from i18next.com

# Library related to the feature
Highcharts

# Proof of Concept/ Live example for the feature
I was trying to make a workaround for this issue, by providing a deep proxy as `lang` property, unfortunately Highcharts is doing deep copy of lang object, which makes it impossible to detect reads for `lang` object properties.

Thanks for understanding! and keep up the good work on Highcharts!

**Comment by sebastianbochan** [comment]
Hi @gtanczyk,
Thank you for the idea!

Regarding to your details:
1. All new features and updates are available in our change log [here](https://www.highcharts.com/blog/changelog/)
2. Im not sure if its realted with Highcharts. Let me explain why. In the setOptions you declare which elements are translated, so when you translate all of them, why you need to detect which are translated in your app?

**Comment by gtanczyk** [comment]
It's all about the cost of translation, and maintainability. I would like spend resources only on those translation keys, which are actually in use inside our application, because we have to support many languages, and it's better to not translate strings, which are never in use. We are not using all of Highcharts features.

I agree that change log is helpful, but with translation callback I can actually automate the process. If I detect a new string, which is not translated and is actually used by our application, I will probably go to change log and docs, to understand the meaning.

I was also trying to get full export of all translation keys from Highcharts, and it's not really possible. You need to construct the `lang` object more or lessby hand. Manually applying changes from change log does not sound appealing in long term.

**Comment by pawelfus** [comment]
Perhaps part of this can be added to the debugger-mode, @sebastianbochan ?

At this moment you can get all keys for lang options: `const defaultLang = Highcharts.getOptions().lang` and compare them with options you have defined already. If `defaultLang` has more properties than your object, you know something was added. This can be added to our debugger-mode. Additionally, you can create (manually of course) list of excluded options that you don't use and prevent from throwing an error. Everything could be setup as a test on your end, preventing from Highcharts upgrade when new lang options were created and were not resolved on your side (by either, translating or excluding).

I'm against the other request: I don't think we should dive into the options you set and inform you which are used or not. That would require checking these values on every `chart.redraw`. Highcharts charts are dynamic, you can start with an empty chart and add elements on the fly (multiple different series types, axes, legend, points etc). We don't know what you are going to use on the chart. Checking these on every `chart.redraw` would decrease performance, which we should definitely avoid.

Any thoughts @oysteinmoseng ? Our a11y already supports i18n, can we somehow improve that?

**Comment by oysteinmoseng** [comment]
The accessibility module uses the `chart.langFormat` function to get the internationalized strings. By wrapping this function you would be able to detect when Highcharts attempts to get internationalized content.

One option could be to introduce a similar concept for the other `lang` keys as well, and consider if we should emit an overridable event in this function. If an event is emitted every time we request a `lang` key, it would be easy to detect which keys are used by listening to the event, as requested by @gtanczyk.

@TorsteinHonsi thoughts?

**Comment by gtanczyk** [comment]
> At this moment you can get all keys for lang options: const defaultLang = Highcharts.getOptions().lang and compare them with options you have defined already

I don't get most of translation keys:
```
{"loading":"Loading...","months":["January","February","March","April","May","June","July","August","September","October","November","December"],"shortMonths":["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"],"weekdays":["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"],"decimalPoint":".","numericSymbols":["k","M","G","T","P","E"],"resetZoom":"Reset zoom","resetZoomTitle":"Reset zoom level 1:1","thousandsSep":" ","viewFullscreen":"View in full screen","printChart":"Print chart","downloadPNG":"Download PNG image","downloadJPEG":"Download JPEG image","downloadPDF":"Download PDF document","downloadSVG":"Download SVG vector image","contextButtonTitle":"Chart context menu","downloadCSV":"Download CSV","downloadXLS":"Download XLS","openInCloud":"Open in Highcharts Cloud","viewData":"View data table"}
```

JSFiddle: https://jsfiddle.net/yLrf5xdn/

**Comment by gtanczyk** [comment]
> The accessibility module uses the chart.langFormat function to get the internationalized strings. By wrapping this function you would be able to detect when Highcharts attempts to get internationalized content.

When I was trying to apply this advice, I realized that I'm not using the accessibility module. The same observation about some Highstock related strings. My translation map went down to two dozens.

**Comment by pawelfus** [comment]
> Using this method I get only small set of translation keys:

That's correct, only a small subset of keys is available in the core. You need to add modules that actually use other options, for example exporting module: https://jsfiddle.net/BlackLabel/asdvbct8/

**Comment by pawellysy** [comment]
unstale
